gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -0,0 +1,251 @@
+            propagate_neighbors=[0, 8, 16],
+            evaluate_neighbors=[9, 9, 9],
+        )
+        model = nn.DataParallel(model)
+
+        # Check if cuda devices is supported, and load the pretrained model
+        #   the pretrained checkpoint should be pre-downloaded using gtsfm/download_model_weights.sh
+        if torch.cuda.is_available():
+            model.cuda()
+            state_dict = torch.load(PATCHMATCHNET_WEIGHTS_PATH)
+        else:
+            state_dict = torch.load(PATCHMATCHNET_WEIGHTS_PATH, map_location=torch.device("cpu"))
+
+        model.load_state_dict(state_dict["model"])
+        model.eval()
+
+        depth_est_list = {}
+        confidence_est_list = {}
+        with torch.no_grad():
+            for batch_idx, sample in enumerate(loader):
+                start_time = time.time()
+
+                # Check if cuda devices is supported, and store the inference data to the target device
+                if torch.cuda.is_available():
+                    sample_device = tocuda(sample)
+                else:
+                    sample_device = copy.copy(sample)
+
+                # Inference using PatchmatchNet
+                outputs = model(
+                    sample_device["imgs"],
+                    sample_device["proj_matrices"],
+                    sample_device["depth_min"],
+                    sample_device["depth_max"],
+                )
+
+                outputs = tensor2numpy(outputs)
+                del sample_device
+
+                ids = sample["idx"]
+
+                # Save depth maps and confidence maps
+                for idx, depth_est, photometric_confidence in zip(
+                    ids, outputs["refined_depth"]["stage_0"], outputs["photometric_confidence"]
+                ):
+
+                    idx = idx.cpu().numpy().tolist()
+                    depth_est_list[idx] = depth_est.copy()
+                    confidence_est_list[idx] = photometric_confidence.copy()
+
+                logger.info(
+                    "[Densify::PatchMatchNet] Iter {}/{}, time = {:.3f}".format(
+                        batch_idx + 1, len(loader), time.time() - start_time
+                    )
+                )
+
+        # Filter inference result with thresholds
+        dense_point_cloud = self.filter_depth(
+            dataset=dataset,
+            depth_list=depth_est_list,
+            confidence_list=confidence_est_list,
+            geo_pixel_thresh=thresholds[0],
+            geo_depth_thresh=thresholds[1],
+            photo_thresh=thresholds[2],
+        )
+
+        return dense_point_cloud
+
+    def filter_depth(
+        self,
+        dataset: PatchmatchNetData,
+        depth_list: Dict[int, np.ndarray],
+        confidence_list: Dict[int, np.ndarray],
+        geo_pixel_thresh: float,
+        geo_depth_thresh: float,
+        photo_thresh: float,
+    ) -> np.ndarray:
+        """Filter depth map and get filtered dense point cloud
```

Comment:
what do we filter based on? I think some more details are needed about the criteria / methodology.

Reply:
PatchmatchNet only produces the depth maps and the confidence maps of each view. Some points should be filtered out because:
1. too low confidence (not fit the RGB image well), filtered by `photo_thresh`
2. cannot keep a consistent geometry among depth maps of ref views and source views, filtered by `geo_*_thresh`

Reply:
Cool, how about we say something like:
```python
"""Create a dense point cloud by filtering depth maps based on estimated confidence maps and consistent geometry.

Consistent geometry is defined as X% of Y metric matching as ABC
"""
```